### ISSUE TITLE
CLN remove exception catching in string_to_dts

### ISSUE THREAD

**Issue by MarcoGorelli** [pandas-dev/pandas#50515]
The `except? -1` can be removed

https://github.com/pandas-dev/pandas/blob/e2598a10db59769ece2f3864ac477f52f28c2850/pandas/_libs/tslibs/np_datetime.pyx#L274-L283

as `string_to_dts` is only ever called with `want_exc=False`

This function would need amending to raise `ValueError` if `string_to_dts` failed

https://github.com/pandas-dev/pandas/blob/502919e4b18e5d5aca584d4ebb84d663fcefffc9/pandas/_libs/tslib.pyx#L79-L100

**Comment by ramvikrams** [comment]
Just wanted to clear a confusion we have to add the raise the `ValueError` in the else block

**Comment by MarcoGorelli** [comment]
Hey - I'm not 100% what would be best yet, I'll get back to you on this

**Comment by ShashwatAgrawal20** [comment]
@MarcoGorelli any updates ?

**Comment by ShashwatAgrawal20** [comment]
take

**Comment by MarcoGorelli** [comment]
I'll give some details later this week (though if you want to start investigating, feel free to go ahead!)

**Comment by ShashwatAgrawal20** [comment]
> I'll give some details later this week (though if you want to start investigating, feel free to go ahead!)

The `string_to_dts function` returns a special error value `'(-1)'` as if an error occurs, but the `_test_parse_iso8601` isn't checking for it, we can simply check the return value of  `string_to_dts` if it's -1 we can raise a `ValueError` 

Any thoughts?

**Comment by ShashwatAgrawal20** [comment]
@MarcoGorelli any updates?

**Comment by ShashwatAgrawal20** [comment]
> do you want to see if it's possible to always have `want_exc` as `True`, and to use a try-except instead of checking the value of `string_to_dts_failed`? This might give a more informative error message

Like this in `'_test_parse_iso8601'`? & setting the `want_exc` to `'True'` in `'string_to_dts'`
```python 
obj = _TSObject()

try:
    string_to_dts(ts, &obj.dts, &out_bestunit, &out_local, &out_tzoffset)
except Exception as e:
    raise ValueError(f"Error parsing timestamp: {ts}. Reason: {e}. Make sure the timestamp is in the correct format.")
```

**Comment by MarcoGorelli** [comment]
there shouldn't be any need to try-except in `_test_parse_iso8601` (it already has `want_exc` as `True`), only in other places which call `string_to_dts` but with `want_exc` as `False`, such as 

https://github.com/pandas-dev/pandas/blob/9a5702c51e2d0440255b44352f0c5f6481e2fa6d/pandas/_libs/tslibs/strptime.pyx#L311-L329

**Comment by ShashwatAgrawal20** [comment]
> there shouldn't be any need to try-except in `_test_parse_iso8601` (it already has `want_exc` as `True`), only in other places which call `string_to_dts` but with `want_exc` as `False`, such as
> 
> https://github.com/pandas-dev/pandas/blob/9a5702c51e2d0440255b44352f0c5f6481e2fa6d/pandas/_libs/tslibs/strptime.pyx#L311-L329

Ohh! A wrap-up call? 
```python 
try:
    string_to_dts(val, &dts, &out_bestunit, &out_local, &out_tzoffset, False, fmt, exact)
    value = npy_datetimestruct_to_datetime(NPY_FR_ns, &dts)
    if out_local == 1:
        tz = timezone(timedelta(minutes=out_tzoffset))
        result_timezone[i] = tz
        out_local = 0
        out_tzoffset = 0
    iresult[i] = value
    check_dts_bounds(&dts)
    continue
except Exception as e:
    raise ValueError(f"Error parsing timestamp: {val}. Make sure the timestamp is in the correct format. Error message: {e}")

```

**Comment by MarcoGorelli** [comment]
it still needs to continue if it failed, so more like
```
try:
    string_to_dts(...)
except:
    pass
else:
    value = npy_datetimestruct_to_datetime(NPY_FR_ns, &dts)
    if out_local == 1:
        tz = timezone(timedelta(minutes=out_tzoffset))
        result_timezone[i] = tz
        out_local = 0
        out_tzoffset = 0
    iresult[i] = value
    check_dts_bounds(&dts)
    continue    
```

**Comment by ShashwatAgrawal20** [comment]
Got it, should be something like this if I am not wrong. 
```python 
try:
    string_to_dts(val, &dts, &out_bestunit, &out_local, &out_tzoffset, False, fmt, exact)
except ValueError as e:
    #pass
    raise ValueError(f"Error parsing timestamp: {val}. Make sure the timestamp is in the correct format. Error message: {e}")
else:
    value = npy_datetimestruct_to_datetime(NPY_FR_ns, &dts)
    if out_local == 1:
        tz = timezone(timedelta(minutes=out_tzoffset))
        result_timezone[i] = tz
        out_local = 0
        out_tzoffset = 0
    iresult[i] = value
    check_dts_bounds(&dts)
    continue
```

**Comment by MarcoGorelli** [comment]
not really, you need to keep the `pass`, as I showed in https://github.com/pandas-dev/pandas/issues/50515#issuecomment-1378486138

**Comment by jbrockmendel** [comment]
> This function would need amending to raise ValueError if string_to_dts failed

Seems like it is simpler to leave string_to_dts as it is than re-introduce bespoke error handling here?

**Comment by MarcoGorelli** [comment]
the only function which would need amending would be `_test_parse_iso8601`, which is only used in tests - outside of tests, `string_to_dts` never raises, so the `except? -1` is probably more complex than it needs to be

**Comment by jbrockmendel** [comment]
i tried this a couple weeks ago and broke a bunch of tests (though not locally IIRC).  not really clear to me what the upside is here.

**Comment by MarcoGorelli** [comment]
might as well close then, thanks!
